code/_onclick/MouseDrag.dm

```diff
@@ -1,16 +1,23 @@
 //If we intercept it return true else return false
-/atom/proc/RelayMouseDrag(src_object, over_object, src_location, over_location, src_control, over_control, params, var/mob/user)
+/atom/proc/RelayMouseDrag(atom/src_object, atom/over_object, src_location, over_location, src_control, over_control, params, var/mob/user)
 	return FALSE
 
-/mob/proc/OnMouseDrag(src_object, over_object, src_location, over_location, src_control, over_control, params)
+/mob/proc/OnMouseDrag(atom/src_object, atom/over_object, src_location, over_location, src_control, over_control, params)
 	if(istype(loc, /atom))
 		var/atom/A = loc
 		if(A.RelayMouseDrag(src_object, over_object, src_location, over_location, src_control, over_control, params, src))
 			return
 
-	if(over_object)
-		if(!incapacitated())
-			var/obj/item/gun/gun = get_active_hand()
-			if(istype(gun) && gun.can_autofire())
-				set_dir(get_dir(src, over_object))
-				gun.Fire(get_turf(over_object), src, params, (get_dist(over_object, src) <= 1), FALSE)
\ No newline at end of file
+	var/obj/item/gun/gun = get_active_hand()
+	if(a_intent == I_HURT && istype(over_object) && (isturf(over_object) || isturf(over_object.loc)) && !incapacitated() && istype(gun))
+		gun.set_autofire(over_object, src)
+
+/mob/proc/OnMouseDown(atom/object, location, control, params)
+	var/obj/item/gun/gun = get_active_hand()
+	if(a_intent == I_HURT && istype(object) && (isturf(object) || isturf(object.loc)) && !incapacitated() && istype(gun))
+		gun.set_autofire(object, src)
+
+/mob/proc/OnMouseUp(atom/object, location, control, params)
```

Comment:
I'm not technically competent enough to offer meaningful feedback here (which is why I was avoiding this PR) but I believe you need to do similar stuff in `client/Destroy`, in the mob client transfer helper, and possibly check stuff in `Life` (as client disconnection doesn't reliably trigger anything). If you already do this, or you feel it's not necessary, then can discuss.

Reply:
As discussed on Discord, the autofire process checks for the mob being destroyed, the client being null, or the weapon not being in the firer's contents, and breaks out early if so.